### PR TITLE
CDAP-8338 Support change system log levels

### PR DESCRIPTION
**Pull request by yaojiefeng**
JIRA: https://issues.cask.co/browse/CDAP-8338
Build: https://builds.cask.co/browse/CDAP-DUT5686-2

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java [hunk 1]
```diff
@@ -153,4 +157,16 @@ protected boolean respondIfElementNotFound(Throwable t, HttpResponder responder)
     }
     return false;
   }
+
+  /**
+   * Helper method to transform the type of the log level map.
+   */
+  protected Map<String, LogEntry.Level> transformLogLevelsMap(Map<String, String> logLevels) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java:165]: Better to transform using `Maps.transformValues`

I noticed you have used `Maps.tranformValues` in `AbstractDistributedProgramRunner` for a similar log level transformation.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java [hunk 2]
```diff
@@ -128,4 +131,14 @@ public void restartAllInstances() {
   public void restartInstances(int instanceId, int... moreInstanceIds) {
     // no-op
   }
+
+  @Override
+  public void updateServiceLogLevels(Map<String, LogEntry.Level> logLevels) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java]: so we don't want to allow changing appfabric log levels, why ? its on edge node, but we could still change it right ?

File: cdap-common/src/main/java/co/cask/cdap/common/twill/MasterServiceManager.java [hunk 3]
```diff
@@ -99,5 +103,17 @@
...
+
+  /**
+   * Update log levels for this service
+   * @param logLevels The {@link Map} contains the requested logger name and log level.
+   */
+  void updateServiceLogLevels(Map<String, LogEntry.Level> logLevels);
+
+  /**
+   * Reset the log levels of the service.
+   * @param loggerNames The set of logger names to be reset, if empty, all log levels will be reset.
```
Comment (gsps1) [inline cdap-common/src/main/java/co/cask/cdap/common/twill/MasterServiceManager.java]: what would reset mean ? would it be reset to some default log level ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java [hunk 4]
```diff
@@ -1656,6 +1655,8 @@ private void updateLogLevels(HttpRequest request, HttpResponder responder, Strin
       responder.sendStatus(HttpResponseStatus.OK);
     } catch (JsonSyntaxException e) {
       throw new BadRequestException("Invalid JSON in body");
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(e.getMessage());
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:1703]: why is this change needed ?

**Review by gsps1 — COMMENTED** [review]
If we are not going to support log level change for AppFabric, we should surface that with better "not-supported" message rather than it being a no-op. however I think we should support changing log levels for app-fabric too.

File: cdap-common/src/main/java/co/cask/cdap/common/twill/MasterServiceManager.java (hunk 3 above)
Comment (yaojiefeng) [reply cdap-common/src/main/java/co/cask/cdap/common/twill/MasterServiceManager.java]: It will reset the log levels back to when the service is launched.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java (hunk 4 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:1703]: Since `LogEntry.valueOf(String level)` can throw this exception and we want to propagate it as `BadRequestException`

File: cdap-common/src/main/java/co/cask/cdap/common/twill/MasterServiceManager.java (hunk 3 above)
Comment (gsps1) [reply cdap-common/src/main/java/co/cask/cdap/common/twill/MasterServiceManager.java]: will be good to add that info to javadoc.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java [hunk 5]
```diff
@@ -64,7 +69,8 @@
     .registerTypeAdapter(EntityId.class, new EntityIdTypeAdapter())
     .create();
 
-  protected static final java.lang.reflect.Type STRING_MAP_TYPE = new TypeToken<Map<String, String>>() { }.getType();
+  protected static final Type STRING_MAP_TYPE = new TypeToken<Map<String, String>>() { }.getType();
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java]: rename to `MAP_STRING_TYPE`, as you are using `SET_STRING_TYPE` in this class. better to follow one format.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java (hunk 4 above)
Comment (gsps1) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java:1703]: It could also throw `NullPointerException` if the parameter`level` is null.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java [hunk 6]
```diff
@@ -153,4 +159,16 @@ protected boolean respondIfElementNotFound(Throwable t, HttpResponder responder)
...
   }
+
+  /**
+   * Helper method to transform the type of the log level map.
+   */
+  protected Map<String, LogEntry.Level> transformLogLevelsMap(Map<String, String> logLevels) {
+    return Maps.transformValues(logLevels, new Function<String, LogEntry.Level>() {
+      @Override
+      public LogEntry.Level apply(String input) {
+        return LogEntry.Level.valueOf(input.toUpperCase());
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java]: can `input` be null? I remember earlier, you handled the case where input is null, in that case it will throw NPE now.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java [hunk 7]
```diff
@@ -134,11 +147,60 @@ public void restartInstances(int instanceId, int... moreInstanceIds) {
...
-    // no-op
+    LoggerContext context = getLoggerContext();
+    Iterator<Map.Entry<String, LogEntry.Level>> entryIterator = oldLogLevels.entrySet().iterator();
+    while (entryIterator.hasNext()) {
+      Map.Entry<String, LogEntry.Level> entry = entryIterator.next();
+      String loggerName = entry.getKey();
+      // logger name is empty if we are resetting all loggers.
+      if (loggerNames.isEmpty() || loggerNames.contains(loggerName)) {
+        setLogLevel(context, loggerName, entry.getValue());
+        entryIterator.remove();
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java:188]: why do you have to remove ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java [hunk 8]
```diff
@@ -134,11 +147,60 @@ public void restartInstances(int instanceId, int... moreInstanceIds) {
...
   @Override
   public void updateServiceLogLevels(Map<String, LogEntry.Level> logLevels) {
-    // no-op
+    LoggerContext context = getLoggerContext();
+    if (context != null) {
+      for (Map.Entry<String, LogEntry.Level> entry : logLevels.entrySet()) {
+        String loggerName = entry.getKey();
+        LogEntry.Level oldLogLevel = setLogLevel(context, loggerName, entry.getValue());
+        // if logger name is not in oldLogLevels, we need to record its original log level
+        if (!oldLogLevels.containsKey(loggerName)) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java:163]: is this logic similar to how we change log levels for runnable ?

**Review by gsps1 — COMMENTED** [review]
few comments

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java (hunk 7 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java:188]: Because after resetting, the log level will be the same as it is in the logback.xml. There is no need to keep it in memory.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java (hunk 8 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java:163]: Yes, this is the same logic as it is in Twill

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 9]
```diff
@@ -268,4 +271,63 @@ public void getLatestRestartServiceInstanceStatus(HttpRequest request, HttpRespo
                            String.format("No restart instances request found or %s", serviceName));
     }
   }
+
+  public void updateServiceLogLevels(HttpRequest request, HttpResponder responder, String serviceName) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: Javadoc for public method.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 10]
```diff
@@ -268,4 +271,63 @@ public void getLatestRestartServiceInstanceStatus(HttpRequest request, HttpRespo
...
+    if (!serviceManagementMap.containsKey(serviceName)) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Invalid service name %s", serviceName));
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: Should use `throw new NotFoundException` instead.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 11]
```diff
@@ -268,4 +271,63 @@ public void getLatestRestartServiceInstanceStatus(HttpRequest request, HttpRespo
...
+      return;
+    }
+
+    MasterServiceManager masterServiceManager = serviceManagementMap.get(serviceName);
+    if (!masterServiceManager.isServiceEnabled()) {
+      responder.sendString(HttpResponseStatus.FORBIDDEN,
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: Similar here. Should throw exception. Currently we don't have one for `Forbidden`. Please create one in cdap-common.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 12]
```diff
@@ -268,4 +271,63 @@ public void getLatestRestartServiceInstanceStatus(HttpRequest request, HttpRespo
...
+      return;
+    }
+
+    try {
+      // we are decoding the body to Map<String, String> instead of Map<String, LogEntry.Level> here since Gson will
+      // serialize invalid enum values to null, which is allowed for log level, instead of throw an Exception.
+      masterServiceManager.updateServiceLogLevels(transformLogLevelsMap(decodeArguments(request)));
+      responder.sendStatus(HttpResponseStatus.OK);
+    } catch (IllegalStateException ise) {
+      responder.sendString(HttpResponseStatus.SERVICE_UNAVAILABLE,
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: Should throw `ServiceUnavailableException`

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 13]
```diff
@@ -268,4 +271,63 @@ public void getLatestRestartServiceInstanceStatus(HttpRequest request, HttpRespo
...
+                           String.format("Failed to update log levels for service %s " +
+                                           "because the service may not be ready yet", serviceName));
+    } catch (IllegalArgumentException e) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, e.getMessage());
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: Should throw `BadRequestException`

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 14]
```diff
@@ -268,4 +271,63 @@ public void getLatestRestartServiceInstanceStatus(HttpRequest request, HttpRespo
...
+    } catch (JsonSyntaxException e) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, "Invalid Json in the body");
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: Same as above

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 15]
```diff
@@ -268,4 +271,63 @@ public void getLatestRestartServiceInstanceStatus(HttpRequest request, HttpRespo
...
+    }
+
+  }
+
+  public void resetServiceLogLevels(HttpRequest request, HttpResponder responder, String serviceName) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: javadoc.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java [hunk 16]
```diff
@@ -268,4 +271,63 @@ public void getLatestRestartServiceInstanceStatus(HttpRequest request, HttpRespo
...
+    if (!serviceManagementMap.containsKey(serviceName)) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Invalid service name %s", serviceName));
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java]: All error responses should be done via throwing exception

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java [hunk 17]
```diff
@@ -36,12 +55,21 @@
  * Monitor Handler V3
  */
 @Path(Constants.Gateway.API_VERSION_3)
-public class MonitorHandler extends AbstractMonitorHandler {
+public class MonitorHandler extends AbstractAppFabricHttpHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MonitorHandler.class);
+
+  private final Map<String, MasterServiceManager> serviceManagementMap;
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/MonitorHandler.java]: group final and static final fields.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java [hunk 18]
```diff
@@ -153,4 +158,16 @@ protected boolean respondIfElementNotFound(Throwable t, HttpResponder responder)
     }
     return false;
   }
+
+  /**
+   * Helper method to transform the type of the log level map.
+   */
+  protected Map<String, LogEntry.Level> transformLogLevelsMap(Map<String, String> logLevels) {
+    return Maps.transformValues(logLevels, new Function<String, LogEntry.Level>() {
+      @Override
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java:167]: add @nullable annotation

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java [hunk 19]
```diff
@@ -128,4 +144,63 @@ public void restartAllInstances() {
...
+          oldLogLevels.put(loggerName, oldLogLevel);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void resetServiceLogLevels(Set<String> loggerNames) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java]: need to add javadoc for public methods

**Review by gsps1 — COMMENTED** [review]
minor comments

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java (hunk 19 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java]: This is a method which is overridden from MasterServiceManager, the javadoc can be found there.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java [hunk 20]
```diff
@@ -145,6 +145,13 @@ public void restartInstances(int instanceId, int... moreInstanceIds) {
     // no-op
   }
 
+  /**
+   * Update log levels for app fabric service.
+   * Note that changing the log levels of app fabric service will also change log levels for cdap master
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java]: for `app.fabric server and dataset service` right ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java [hunk 21]
```diff
@@ -160,6 +167,14 @@ public void updateServiceLogLevels(Map<String, LogEntry.Level> logLevels) {
     }
   }
 
+  /**
+   * Reset the log levels of app fabric service.
+   * The log levels will be reset back to when the service was launched.
+   * Note that resetting the log levels of app fabric service will also reset log levels for cdap master
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java]: same here

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java [hunk 22]
```diff
@@ -160,6 +167,14 @@ public void updateServiceLogLevels(Map<String, LogEntry.Level> logLevels) {
...
+   * and dataset.service.
+   *
+   * @param loggerNames The set of logger names to be reset, if empty, all log levels will be reset.
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java]: all log levels ? or all loggers ?  should it be `all loggers will be reset to the level when the service started`

**Review by gsps1 — COMMENTED** [review]
doc comments

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java (hunk 20 above)
Comment (gsps1) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AppFabricServiceManager.java]: not master.

**Comment by gsps1** [comment]
LGTM, squash and merge.

**Review by gsps1 — APPROVED** [review]
